### PR TITLE
feat(swarm): don't have `ConnectionHandler`s close connections

### LINKED ISSUE #3591: swarm: Remove `ConnectionHandler::Error`
Based on https://github.com/libp2p/rust-libp2p/discussions/3353#discussioncomment-4733941, I am opening this issue to track the work around removing `ConnectionHandlerEvent::Close`. Closing a connection is the only usage of `ConnectionHandler::Error` but we've agree that `ConnectionHandler`s themselves should not be allowed to close connections because they don't fully own them and might disrupt other protocols.

```[tasklist]
- [x] https://github.com/libp2p/rust-libp2p/pull/3625
- [x] https://github.com/libp2p/rust-libp2p/pull/3913
- [x] https://github.com/libp2p/rust-libp2p/pull/3947
- [ ] https://github.com/libp2p/rust-libp2p/pull/4714
- [ ] https://github.com/libp2p/rust-libp2p/pull/4715
- [ ] https://github.com/libp2p/rust-libp2p/pull/4718
- [ ] https://github.com/libp2p/rust-libp2p/pull/4745
- [ ] https://github.com/libp2p/rust-libp2p/pull/4749
- [ ] https://github.com/libp2p/rust-libp2p/pull/4755
```

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

This PR implements the long-awaited design of disallowing `ConnectionHandler`s to close entire connections. Instead, users should close connections via `ToSwarm::CloseConnection` from a `NetworkBehaviour` or - even better - from the `Swarm` via `close_connection`. A `NetworkBehaviour` also does not have a "full" view onto how a connection is used but at least it can correlate whether it created the connection via the `ConnectionId`. In general, the more modular and friendly approach is to stop "using" a connection if a particular protocol no longer needs it. As a result of the keep-alive algorithm, such a connection is then closed automatically.

Depends-on: #4745.
Depends-on: #4718.
Depends-on: #4749.
Related: #3353.
Related: #4714.
Resolves: #3591.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by mxinden** [comment]
Let me know once this is ready for a review. Direction looking good.

**Comment by thomaseizinger** [comment]
> Let me know once this is ready for a review. Direction looking good.

Needs all linked PRs to merge first before the build is green, otherwise ready for review :)

**Comment by thomaseizinger** [comment]
@mxinden All dependencies of this PR have been merged so this is ready for review.

File: swarm/CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,9 @@
   See [PR 4076](https://github.com/libp2p/rust-libp2p/pull/4076).
 - Remove deprecated `PollParameters` from `NetworkBehaviour::poll` function.
   See [PR 4490](https://github.com/libp2p/rust-libp2p/pull/4490).
+- Remove deprecated `ConnectionHandlerEvent::Close` and `ConnectionHandler::Error`.
+  `ConnectionHandler`s should not close connections directly as the connection might still be in use by other handlers.
+  See [PR XXXX](https://github.com/libp2p/rust-libp2p/pull/XXXX).
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion
  See [PR 4755](https://github.com/libp2p/rust-libp2p/pull/4755).
```

**Review by mxinden — APPROVED** [review]
Great to see this final step.

**Pull request merged**